hal/src/hal-sdcard.ads

```diff
@@ -0,0 +1,426 @@
+   SD_App_Get_CER_RN2                 : constant SD_Specific_Command := 46;
+   SD_App_Set_CER_RES2                : constant SD_Specific_Command := 47;
+   SD_App_Get_CER_RES1                : constant SD_Specific_Command := 48;
+   SD_App_Change_Secure_Area          : constant SD_Specific_Command := 49;
+   SD_App_Send_SCR                    : constant SD_Specific_Command := 51;
+
+   type Rsp_Kind is
+     (Rsp_Invalid,
+      --  Unknown/invalid command
+
+      Rsp_No,
+      --  No response.
+
+      Rsp_R1, Rsp_R1B, Rsp_R2, Rsp_R3, Rsp_R6, Rsp_R7
+      --  Response with format
+     );
+   --  Response format
+
+   type Tfr_Kind is
+     (Tfr_Invalid, Tfr_No,
+      Tfr_Read, Tfr_Write,
+      Tfr_Read_Multi, Tfr_Write_Multi);
+   --  Transfer
+
+   type Cmd_Desc_Type is record
+      Cmd : SD_Command;
+      Rsp : Rsp_Kind;
+      Tfr : Tfr_Kind;
+   end record;
+
+   type SDCard_Driver is limited interface;
+
+   procedure Reset
+     (This   : in out SDCard_Driver;
+      Status : out SD_Error) is abstract;
+   --  Initialize the driver, enable clocking.
+
+   procedure Set_Clock
+     (This : in out SDCard_Driver;
+      Freq : Natural) is abstract;
+   --  Set clock frequency.
+
+   procedure Set_Bus_Size
+     (This : in out SDCard_Driver;
+      Mode : Wide_Bus_Mode) is abstract;
+   --  Set host bus size; the command must have been set to the card.
+
+   procedure Send_Cmd
+     (This   : in out SDCard_Driver;
+      Cmd    : Cmd_Desc_Type;
+      Arg    : Unsigned_32;
+      Status : out SD_Error) is abstract;
+   --  Send a command (without data transfer) and wait for result.
+
+   procedure Send_Cmd
+     (This   : in out SDCard_Driver'Class;
+      Cmd    : SD_Command;
+      Arg    : Unsigned_32;
+      Status : out SD_Error);
+   --  Wrapper for Send_Cmd using a generic command.
+
+   procedure Send_ACmd
+     (This   : in out SDCard_Driver'Class;
+      Cmd    : SD_Specific_Command;
+      Rca    : Unsigned_16;
+      Arg    : Unsigned_32;
+      Status : out SD_Error);
+   --  Send application specific command
+
+   procedure Read_Cmd
+     (This   : in out SDCard_Driver;
+      Cmd    : Cmd_Desc_Type;
+      Arg    : Unsigned_32;
+      Buf    : System.Address;
+      Len    : Unsigned_32;
+      Status : out SD_Error) is abstract;
+   --  Read data command
+
+   procedure Read_Rsp48
+     (This : in out SDCard_Driver;
+      Rsp  : out Unsigned_32) is abstract;
+   --  Read the 32 interesting bits of the last 48bits response (start bit,
+   --  transmission bit, command index, crc and end bit are discarded).
+   --  Cannot fail.
+
+   procedure Read_Rsp136
+     (This           : in out SDCard_Driver;
+      W0, W1, W2, W3 : out Unsigned_32) is abstract;
+   --  Read the 128 interesting bits of the last 136 bit response.
+   --  W0 is the MSB, W3 the LSB
+   --  Cannot fail.
+
+   procedure Card_Identification_Process (Driver : in out SDCard_Driver'Class;
```

Comment:
So you are saying that Implementation of the SDMMC software protocol should not be in the same library as file systems. Fine, if you want to create a yet another library project to put the just SDMMC software protocol that's not a huge deal. I don't think it's necessary. My point is that this code should not be in the HAL because it's not an abstraction it's an implementation.

I did the separation for you in #116, there's no circular dependency and I also put back the Cortex-M0 support that was removed for no reason.